force-app/main/default/lwc/todoList/__tests__/todoList.test.js

```diff
@@ -0,0 +1,108 @@
+import { createElement } from 'lwc';
+import TodoList from 'c/todoList';
+
+const todos = [
+    { id: 1, description: 'Explore recipes', priority: true },
+    { id: 2, description: 'Install Ebikes sample app', priority: false }
+];
+
+describe('c-todo-list', () => {
+    afterEach(() => {
+        // The jsdom instance is shared across test cases in a single file so reset the DOM
+        while (document.body.firstChild) {
+            document.body.removeChild(document.body.firstChild);
+        }
+    });
+
+    it('renders one lightning-input element and no list items as default', () => {
+        // Create initial element
+        const element = createElement('c-todo-list', {
+            is: TodoList
+        });
+        document.body.appendChild(element);
+
+        // Query for lightning-input elemenst
+        const lightningInputEl = element.shadowRoot.querySelector(
+            'lightning-input'
+        );
+        expect(lightningInputEl).not.toBeNull();
+        expect(lightningInputEl.label).toBe('Priority Only');
+        expect(lightningInputEl.checked).toBeFalsy();
+
+        // Query for rendered list items
+        const listItemEls = element.shadowRoot.querySelectorAll('li');
+        expect(listItemEls.length).toBe(0);
+
+        // Return a promise to wait for any asynchronous DOM updates. Jest
+        // will automatically wait for the Promise chain to complete before
+        // ending the test and fail the test if the promise ends in the
+        // rejected state
+        return Promise.resolve().then(() => {
+            // Compare if tracked property has been assigned a new value.
+        });
+    });
+
+    it('renders multiple list items and filters based on priority', () => {
```

Comment:
In general, we should only be testing a single behavior per test case. I'd rather see this as two separate test cases. One that verifies you can push an item to the `todos` list and it gets rendered to the list. Another case that just takes the hardcoded list of todos and verifies clicking the filter checkbox filters out one of the todos.

Reply:
Changed